examples/solids/index.rst

```diff
@@ -142,12 +142,12 @@ Consequently, equation :math:numref:`derss` can be written in matrix form as fol
    \right]  = 
    \left[
      \begin{array}{cccccc} 
-       2\mu +\bar{\lambda} & \bar{\lambda} & \bar{\lambda} & & & \\
-        \bar{\lambda} & 2\mu +\bar{\lambda} & \bar{\lambda} & & & \\
-        \bar{\lambda} & \bar{\lambda} & 2\mu +\bar{\lambda} & & & \\
-        & & & \mu & & \\    
-        & & & &\mu & \\
-        & & & & & \mu \\   
+       2 \mu +\bar{\lambda} & \bar{\lambda} & \bar{\lambda} & & & \\
+       \bar{\lambda} & 2 \mu +\bar{\lambda} & \bar{\lambda} & & & \\
+       \bar{\lambda} & \bar{\lambda} & 2 \mu +\bar{\lambda} & & & \\
+       & & & 2 \mu & & \\    
+       & & & & 2 \mu & \\
+       & & & & & 2 \mu \\   
```

Comment:
Note that the original was correct if you use the scaling of Voigt notation. If you're just ordering without the scaling convention (which won't match textbooks), then the matrices in the linear elasticity section need to also have their scaling updated. This equation should also use `\begin{pmatrix}`. Cc: @ArashMehraban

Reply:
@jeremylt  I think this order:
d\sigma_11 , d\sigma_22, d\sigma_33, d\sigma_12, d\sigma_13, d\sigma_23
should be changed to:
 d\sigma_11 , d\sigma_22, d\sigma_33, d\sigma_23, d\sigma_13, d\sigma_12

Reply:
This comment and diagram may be useful, and the narrative could be adjusted to introduce it earlier.
![Screenshot_20200323_222831](https://user-images.githubusercontent.com/3303/77388415-a7051f80-6d55-11ea-9535-60c77be70f44.png)

I'll let one of you repair the currently-inconsistent Voigt-like scaling between the linear and small-strain sections.

Reply:
I expanded the numerical stability note, and this will allow removing some code in `hyperFS.h`.
![Screenshot_20200324_103435](https://user-images.githubusercontent.com/3303/77452050-16f9c100-6dbb-11ea-88a2-635b4abc394f.png)

Reply:
To make the point about numerical stability, consider this demo. If you choose single precision, neither of these quantities will have the first digit correct. `H` here is meant to represent `\nabla_X u`.
![Screenshot_20200324_100501](https://user-images.githubusercontent.com/3303/77453303-b66b8380-6dbc-11ea-9fc2-34e9c8484840.png)

Reply:
Updated - on my machine this is faster for larger nu and slower for smaller nu, compared to the previous formulation.

Reply:
"slower"? Perhaps a Newton bug? I'll comment on this patch.

Reply:
Well, I deleted some unnecessary code. We're getting poor vectorization of the quadrature loop, but the code looks good to me. Can you give me two cases to compare to see slower for small nu?

Reply:
It's better than it was, so something must have been going in the background on my machine.

There is now a small consistent performance difference on the 672 element mesh (~3-5 more seconds), but I think that's because on this mesh I'm getting the same number of SNES iterations and what looks to the about the same number of KSP iterations, but with slightly more computations in the QFunction. I don't seem to be in a region where the extra stability is helping cut iterations with higher accuracy.

Reply:
This choice isn't about cutting iterations for these large-deformation problems, but rather being able to solve small-deformation problems to high relative accuracy despite using the FS formulation. Our vectorization is held up entirely because `log1p` from `libc` isn't vectorized. We get decently vectorized code if we replace it with a local approximation (and almost double the end-to-end performance).
```
00000000000131b9 <HyperFSdF+0xcc9> vfmadd231pd ymm9,ymm1,ymm3
00000000000131be <HyperFSdF+0xcce> vfmadd231pd ymm14,ymm1,ymm2
00000000000131c3 <HyperFSdF+0xcd3> vfmadd231pd ymm9,ymm15,ymm4
00000000000131c8 <HyperFSdF+0xcd8> vfmadd231pd ymm14,ymm15,ymm6
00000000000131cd <HyperFSdF+0xcdd> vmovapd YMMWORD PTR [rsp+0x288],ymm9
00000000000131d6 <HyperFSdF+0xce6> vmulpd ymm9,ymm13,ymm7
00000000000131da <HyperFSdF+0xcea> vmulpd ymm13,ymm13,ymm4
00000000000131de <HyperFSdF+0xcee> vfmadd231pd ymm9,ymm0,ymm3
00000000000131e3 <HyperFSdF+0xcf3> vfmadd231pd ymm13,ymm0,ymm2
00000000000131e8 <HyperFSdF+0xcf8> vmovapd ymm0,YMMWORD PTR [rsp+0x1e8]
00000000000131f1 <HyperFSdF+0xd01> vmulpd ymm0,ymm0,YMMWORD PTR [rsp-0x58]
00000000000131f7 <HyperFSdF+0xd07> vfmadd231pd ymm9,ymm12,ymm4
00000000000131fc <HyperFSdF+0xd0c> vfmadd231pd ymm13,ymm12,ymm6
0000000000013201 <HyperFSdF+0xd11> vmovapd YMMWORD PTR [rsp+0x228],ymm9
000000000001320a <HyperFSdF+0xd1a> vmulpd ymm9,ymm11,ymm7
000000000001320e <HyperFSdF+0xd1e> vmulpd ymm11,ymm11,ymm4
0000000000013212 <HyperFSdF+0xd22> vfmadd231pd ymm9,ymm5,ymm3
0000000000013217 <HyperFSdF+0xd27> vfmadd231pd ymm11,ymm5,ymm2
000000000001321c <HyperFSdF+0xd2c> vfmadd231pd ymm9,ymm10,ymm4
0000000000013221 <HyperFSdF+0xd31> vfmadd132pd ymm10,ymm11,ymm6
0000000000013226 <HyperFSdF+0xd36> vmovapd ymm11,YMMWORD PTR [rsp+0x408]
000000000001322f <HyperFSdF+0xd3f> vsubpd ymm11,ymm11,YMMWORD PTR [rsp+0x328]
```

We don't have to resolve this now, but [SVML](https://software.intel.com/en-us/cpp-compiler-developer-guide-and-reference-intrinsics-for-logarithmic-operations-512-bit) has a vectorized `log1p` that we could work out how to call. Alternatively, we can rejigger the portable code to help the compiler vectorize despite the fact that _that_ part can't vectorize. That would probably involve some manual tiling.
```diff
diff --git i/examples/solids/qfunctions/hyperFS.h w/examples/solids/qfunctions/hyperFS.h
index 48dd06da..8bb443bc 100644
--- i/examples/solids/qfunctions/hyperFS.h
+++ w/examples/solids/qfunctions/hyperFS.h
@@ -176,7 +176,7 @@ CEED_QFUNCTION(HyperFSF)(void *ctx, CeedInt Q, const CeedScalar *const *in,
     // *INDENT-ON*
 
     // Compute the Second Piola-Kirchhoff (S)
-    const CeedScalar llnj = lambda*log1p(detC_m1)/2.;
+    const CeedScalar llnj = lambda*detC_m1/2.;
     const CeedScalar S00 = llnj*Cinv[0][0] +
                            mu*(Cinv[0][0]*E2[0][0] + Cinv[0][1]*E2[1][0] +
                                Cinv[0][2]*E2[2][0]),
@@ -371,7 +371,7 @@ CEED_QFUNCTION(HyperFSdF)(void *ctx, CeedInt Q, const CeedScalar *const *in,
     // *INDENT-ON*
 
     // Compute the Second Piola-Kirchhoff (S)
-    const CeedScalar llnj = lambda*log1p(detC_m1)/2.;
+    const CeedScalar llnj = lambda*detC_m1/2.;
     const CeedScalar S00 = llnj*Cinv[0][0] +
                            mu*(Cinv[0][0]*E2[0][0] + Cinv[0][1]*E2[1][0] +
                                Cinv[0][2]*E2[2][0]),
```

Reply:
Would a second or third order approximation provide sufficient accuracy and allow for vectorization?

Reply:
Yeah, definitely possible. This series is probably the one to use. The modeling error for any real material will almost certainly be larger than the approximation error.

What's your feeling about factoring utility functions into static inline functions that we call from qfunctions? I don't think there's a portability problem with that and it would allow a lot less duplicated code between the HyperFSF and HyperFSdF.
![Screenshot_20200324_202713](https://user-images.githubusercontent.com/3303/77495322-d8d9bd00-6e0d-11ea-87c6-cd78caf3f159.png)

Reply:
I unrolled the loop in that series and now we have a much faster hyperFS.